Content.Shared/Alert/AlertsSystem.cs

```diff
@@ -71,107 +124,153 @@ public bool TryGetAlertState(EntityUid euid, AlertKey key, out AlertState alertS
+        // Check whether the alert category we want to show is already being displayed, with the same type,
+        // severity, and cooldown.
+        if (entity.Comp.Alerts.TryGetValue(alert.AlertKey, out var alertStateCallback) &&
+            alertStateCallback.Type == alertType &&
+            alertStateCallback.Severity == severity &&
+            alertStateCallback.Cooldown == cooldown &&
+            alertStateCallback.AutoRemove == autoRemove &&
+            alertStateCallback.ShowCooldown == showCooldown)
+        {
+            return;
+        }
 
-            var state = new AlertState
-                { Cooldown = cooldown, Severity = severity, Type = alertType, AutoRemove = autoRemove, ShowCooldown = showCooldown};
-            alertsComponent.Alerts[alert.AlertKey] = state;
+        // In the case we're changing the alert type but not the category, we need to remove it first.
+        entity.Comp.Alerts.Remove(alert.AlertKey);
 
-            // Keeping a list of AutoRemove alerts, so Update() doesn't need to check every alert
-            if (autoRemove)
-            {
-                var autoComp = EnsureComp<AlertAutoRemoveComponent>(euid);
-                if (!autoComp.AlertKeys.Contains(alert.AlertKey))
-                    autoComp.AlertKeys.Add(alert.AlertKey);
-            }
-
-            AfterShowAlert((euid, alertsComponent));
+        var state = new AlertState
+            { Cooldown = cooldown, Severity = severity, Type = alertType, AutoRemove = autoRemove, ShowCooldown = showCooldown};
+        entity.Comp.Alerts[alert.AlertKey] = state;
 
-            Dirty(euid, alertsComponent);
+        // Keeping a list of AutoRemove alerts, so Update() doesn't need to check every alert
+        if (autoRemove)
+        {
+            var autoComp = EnsureComp<AlertAutoRemoveComponent>(entity);
+            if (!autoComp.AlertKeys.Contains(alert.AlertKey))
+                autoComp.AlertKeys.Add(alert.AlertKey);
         }
-        else
+
+        AfterShowAlert((entity, entity.Comp));
+
+        Dirty(entity, entity.Comp);
+    }
+
+    /// <summary>
+    /// An alternative to show alert with different behavior if an alert already exists.
+    /// </summary>
+    /// <param name="entity">Entity whose alert we're updating</param>
+    /// <param name="alertType">Prototype of the alert we're updating</param>
+    /// <param name="severity">Severity we're setting the alert to</param>
+    /// <param name="cooldown">Time left in the alert.</param>
+    /// <param name="autoRemove">Do we want to remove this alert when it expires?</param>
+    /// <param name="showCooldown">Should we show/hide the cooldown?</param>
+    public void UpdateAlert(Entity<AlertsComponent?> entity,
+        ProtoId<AlertPrototype> alertType,
+        short? severity = null,
+        TimeSpan? cooldown = null,
+        bool autoRemove = false,
+        bool showCooldown = true)
+    {
+        if (_timing.ApplyingState)
```

Comment:
Huh... I didn't realize alerts had custom networking stuff. There's probably a few ways we could make it all be auto handled, if we wanted. Most idiomatic would be to eventify alerts (big scope), but I think there's an arguably okay way.

We could add a field to AlertsComponent that's client exclusive (by way of just never touching it on the server, not by comp inheritance) . Then for drawing the alerts the client just uses both sets.

This can all be considered out of scope for this PR, though. :)